ebpf/bpf/pyperf.bpf.c

```diff
@@ -0,0 +1,548 @@
+                &ptr, sizeof(void *), (void *) (cur_frame + offsets->VFrame_localsplus))) {
+            bpf_dbg_printk("failed to read f_localsplus at %x\n", cur_frame + offsets->VFrame_localsplus);
+            return -PY_ERROR_CLASS_NAME;
+        }
+        if (ptr) {
+            if (first_self) {
+                // we are working with an instance, first we need to get type
+                if (bpf_probe_read_user(&ptr, sizeof(void *), ptr + PY_OFFSET_PyObject_ob_type)) {
+                    bpf_dbg_printk("failed to read ob_type at %x\n", ptr + PY_OFFSET_PyObject_ob_type);
+                    return -PY_ERROR_CLASS_NAME;
+                }
+            }
+            if (bpf_probe_read_user(&ptr, sizeof(void *), ptr + PY_OFFSET_PyTypeObject_tp_name)) {
+                bpf_dbg_printk("failed to read tp_name at %x\n", ptr + PY_OFFSET_PyTypeObject_tp_name);
+                return -PY_ERROR_CLASS_NAME;
+            }
+            if (bpf_probe_read_user_str(&symbol->classname, sizeof(symbol->classname), ptr) < 0) {
+                bpf_dbg_printk("failed to read class name at %x\n", ptr);
+                return -PY_ERROR_CLASS_NAME;
+            }
+        } else {
+            // this happens in rideshare flask example under 3.9.18
+            // todo: we should be able to get the class name
+            // https://github.com/fabioz/PyDev.Debugger/blob/2cf10e3fb2ace33b6ef36d66332c82b62815e856/_pydevd_bundle/pydevd_utils.py#L104
+            *((u64 *) symbol->classname) = 0x736c436c6c754e; // NullCls
+        }
+    }
+
+    void *pystr_ptr;
+    // read PyCodeObject's filename into symbol
+    if (bpf_probe_read_user(
+            &pystr_ptr, sizeof(void *), code_ptr + offsets->PyCodeObject_co_filename)) {
+        return -PY_ERROR_FILE_NAME;
+    }
+    if (bpf_probe_read_user_str(
+            &symbol->file, sizeof(symbol->file), pystr_ptr + PY_OFFSET_String_size) < 0) {
+        return -PY_ERROR_FILE_NAME;
+    }
+    // read PyCodeObject's name into symbol
+    if (bpf_probe_read_user(
+            &pystr_ptr, sizeof(void *), code_ptr + offsets->PyCodeObject_co_name)) {
+        return -PY_ERROR_NAME;
+    }
+    if (bpf_probe_read_user_str(
+            &symbol->name, sizeof(symbol->name), pystr_ptr + PY_OFFSET_String_size) < 0) {
+        return -PY_ERROR_NAME;
+    }
+    return 0;
+}
+
+// get_frame_data reads current PyFrameObject filename/name and updates
+// stack_info->frame_ptr with pointer to next PyFrameObject
+// since 311 frame_ptr is pointing to _PyInterpreterFrame
+// returns -PY_ERR_XXX on error, 1 on success, 0 if no more frames
+static __always_inline int get_frame_data(
+        void **frame_ptr,
+        py_offset_config *offsets,
+        py_symbol *symbol,
+        // ctx is only used to call helper to clear symbol, see documentation below
+        void *ctx) {
+    void *code_ptr;
+    void *cur_frame = *frame_ptr;
+    if (!cur_frame) {
+        return 0;
+    }
+
+    //        py311
```

Comment:
Should that be commented ? may be add why.

Reply:
removed